---
title: 'Validating a new computational pipeline for NET-seq data from GSE159603'
author: Robert D. Shear
output: BiocStyle::pdf_document
knit: (function(inputFile, encoding) {
  rmarkdown::render(inputFile, encoding = encoding,
   output_dir = "./knitted", ) })
params:
  case: ["geo", "new"]
  prefix: ["/n/groups/churchman/GSE159603/", 
    "~/Projects/netseq-pipeline/test_results/wt-1_20210829/outputs/"]
  samples: ["wt-1"]
  tileSize: 100
  sampleProportion: 0.1
editor_options: 
  markdown: 
    wrap: 72
---
```{r setup, include=TRUE, message=FALSE}
options(digits = 3)
knitr::opts_chunk$set(message = FALSE)

library(GenomicRanges)
library(GenomicAlignments)
library(rtracklayer)

library(TxDb.Scerevisiae.UCSC.sacCer3.sgdGene)
library(BSgenome.Scerevisiae.UCSC.sacCer3)

library(magick)
library(kableExtra)
library(Gviz)

# Tidyverse declared last...bioconductor homonyms will be masked (e.g. select)
library(tidyverse)
library(glue)
library(plyranges)

# For repeatably
set.seed(20210902)
```

# Introduction

This working paper compares the results of a new NET-seq read-to-occupancy pipeline with published results from the Churchman Lab's NET-seq screen data set.

## Background

The Churchman Lab conducted NET-seq 42 strains of _S. Cerevisiae_ each missing a known elongation factor, recently released in the GEO data series GSE159603, 
"Dynamics of transcription elongation are
finely-tuned by dozens of regulatory factors (baker's yeast)".

I used the wild type controls from this data set for my 2020 thesis, _Inferring High Resolution Transcription Elongation Dynamics from Native Elongating Transcript Sequencing (NET-seq)._
As part of my investigation, I created a revision of the lab's existing sequence read to RNAPII occupancy count (FASTQ-to-bedgraph) pipeline. 

I am now preparing a paper for publication based on this thesis.
In looking back at this software, I decided to investigate modernizing the pipeline.
My modifications took place from 2016 to 2019 and the original software predates this.
There have been several significant improvements in the relivant tools and techniques.

I have created a new NET-seq FASTQ-to-bedgraph pipeline with the following characteristics

- Adoption of Broad Institue published best practices where applicable and practical
- Cromwell scientific workflow manager vs. bash / Snakemake / slurm
- GATK4 vs. custom scripts
- continue use of SAMTOOLS and BEDTOOLS in conjuction with GAtK4 where appropriate
- umi-tools (and possibly others) for UMI processing


## Data Sources

The published results are bedgraph files downloaded from the GEO repository. 
Our trial results were processed on the terra.bio /Google Cloud Platform using the pipeline at [TBS] label [TBS]

We will refer to these two cases as `r paste0(params$case[1], " and ", params$case[2])` respectively.

The data sources used in this run are given in the table below.
We have removed all reads that map to the mitochondrial chomosome (chrM).

## Processing


```{r load.bedgraphs}
strandtab <- c(pos="+", neg="-")
sacCer3Ranges <- SeqinfoForBSGenome("sacCer3")
sacCerRangeNoMito <- dropSeqlevels(sacCer3Ranges, "chrM")

tiles <- unlist(tileGenome(sacCerRangeNoMito, tilewidth = params$tileSize))
if (params$sampleProportion < 1) {
  tiles <- sort(sample(tiles, length(tiles) * params$sampleProportion))
}

# Create ab, the 'score' data for all the cases
# members are the keys sample, case and strand
# bedgraph - The GRranges from the bedgraph file (with mito chromosome removed)
# rleScore - An Integer Rle of occupancy 
#   (one for each chromosome, see GenomicRanges::moclAsRleList for structure)
ab <- crossing(sample = params$samples, caseid = seq_along(params$case), 
               strand = names(strandtab)) %>%   
  mutate(case = params$case[caseid], dir = params$prefix[caseid],
         name = glue_data(., "{case}.{sample}.{strand}"),
        bedgraph = map(
                      map(glue_data(., "{dir}{sample}.{strand}.bedgraph.gz"), 
                        import.bedGraph, genome = "sacCer3"),
                      dropSeqlevels, 
                        "chrM", pruning.mode = "coarse")) %>%
    mutate(occ_tiled = map(.$bedgraph, 
             function(u) (join_overlap_intersect(u, tiles) %>%
              mutate(occ = score * width(.)) %>%
              join_overlap_left(tiles, .) %>%
              group_by(seqnames, start, end)  %>% 
              summarize(occ = sum(occ, na.rm = TRUE), 
                        maxocc = replace_na(max(occ) ,0)) %>%
              as_tibble())
              ),
           bedgraph.rows = map_int(.$bedgraph, length)
  ) %>%
  mutate(mean.density = map_dbl(.$bedgraph, function(u) 
    sum(width(u) * u$score) / sum(seqlengths(sacCerRangeNoMito))))


ab %>% select(sample, strand, case, bedgraph.rows, mean.density) %>%
  kable(caption = "Bedgraph file contents") %>% 
  kable_styling()

```
Note that the mean densities of the _new_ samples are substantially lower than
those of the _geo_ samples. 

Both pipelines intend to remove duplicates, so i would expect the mean densities 
to be almost identical.

## Gross Correspondence 

I divided the genome into `r params$tileSize` nucleotide tiles and 
selected a random sample of `r params$sampleProportion * 100`% of the tiles to
see if how correlated the two cases were at this level.


```{r Examine.Tiles}


o_segs <- ab %>% 
  unnest(occ_tiled) %>% 
  mutate(len = end - start + 1) %>%
  group_by(sample, strand, case, seqnames)

o_tiles <- o_segs %>% 
  summarize(occ = sum(occ, na.rm = TRUE), len = sum(len, na.rm = TRUE)) %>%
  mutate(density = occ / len)

ggplot(o_tiles, aes(x = seqnames, y = density, group = case, col = case)) + 
  geom_point() + facet_wrap(vars(sample, strand), ncol = 1) +
  labs(title = "Mean occupancy / nt by chromosome")

x <- o_tiles %>% 
  pivot_wider(id_cols = c(sample, strand, case), 
              names_from = seqnames, values_from = density)

# HACK  brute force to split the table in two
lapply(list(x[, 1:11], x[, c(1:3, 12:19)]), function(u) {
    kable(u, booktabs = TRUE, caption = "Mean occupancy / nt by chromosome") %>%
    kable_styling() %>%
    print()
})

```

```{r}

dta <- o_segs %>%
  mutate(density = occ / len) %>%
  pivot_wider(id_cols = c(sample, strand, seqnames, start), 
              names_from = case, values_from = density) %>%
  group_by(sample, strand)

dta %>% 
  summarize(tau = cor(geo, new, method = "kendall")) %>% 
  kable() %>% kable_styling()

ggplot(dta, aes(x = geo, y = new)) + 
  geom_point(color = 'blue', alpha = 0.1) + 
  facet_wrap(vars(sample, strand), ncol = 1)  + 
  scale_x_log10() + 
  scale_y_log10()

dtb <- o_segs %>% 
  group_by(sample, strand, case) %>% 
  transmute(sample, strand, case, seqnames, start, density = occ / len)


ggplot(dtb, aes(case, density)) + 
  geom_boxplot(aes(fill = case)) +
  facet_wrap(~sample + strand) +
  scale_y_log10()
```
## What accounts for the density difference?

It looks like it is throughout.

```{r}
q <- ab %>% mutate(q = map(.$bedgraph, 
                      function(u) quantile(rep(u$score, width(u)), c(0,.25,.75,seq(.9,.99,.01),seq(.991,1,.001))))) %>% 
  select(q) %>% 
  unnest_wider(q)

kable(t(q), col.names = ab$name, caption = "Occupancy quantiles") %>% kable_styling()


max_sample_index <- which.max(apply(q, 1, max))
max_score_loc <- (function(u) u[which.max(u$score)])(ab$bedgraph[[max_sample_index]])

```
## Biggest peaks

The highest peak for geo is wildly higher than the higest peaks for new.
The highest peak for wt-1 geo is at `r paste0(ab$name[max_sample_index], "::", max_score_loc)`.

```{r}

# plyranges does not currently (Sep-2021) have full outer join, 
#   so we'll fill in zeros to make the inner join

join_overlap_full_outer <- function(a, b) {
  x <- join_overlap_intersect(a,b)
  y <- join_overlap_intersect(a, complement_ranges(b))
  names(mcols(y)) <- "score.x"
  y$score.y = 0
  z <- join_overlap_intersect(complement_ranges(a), b)
  mcols(z) = list(score.x = rep(0, length(z)), score.y = z$score)
  foo <- sort(unlist(GRangesList(list(x, y, z))))
  }


ab1 <- ab %>%  
  # mutate(bedgraph = map(.$bedgraph, function(u) u[51:75])) %>%
  pivot_wider(id_cols = c(sample, strand), names_from = case, values_from = bedgraph) %>%
  mutate(ov = map2(geo, new, function(u, v) {
              x <- join_overlap_full_outer(u, v)
              x$diff <- x$score.x - x$score.y
              x
            })
     ) %>%
  mutate(nt_occ = map_int(.$ov, function(u) sum(width(u))),
            nt_free = map_int(.$ov, function(u) sum(width(filter(gaps(u), as.character(strand) == '*')))),
            nt_diff = sum(seqlengths(sacCerRangeNoMito)) - nt_occ - nt_free)

# TODO descriptives here
u <- rep(ab1$ov[[1]])
x <- rep(u$diff, width(u))
close.x <- x[abs(x) < 11]
far.x <- x[abs(x) >= 11]

tbl.close <- table(close.x)
print(tbl.close)
print(tbl.close / length(x))


head(sort(far.x, decreasing = TRUE), n = 100)


```

```{r eval=FALSE}
v <- data.frame(new = binned_pos_new$meanScore, old = binned_pos_old$meanScore)
plot(v)
filter <- GRanges("chrIV:437850-437860")
 
old_filtered <- subsetByOverlaps(old_pos, filter)
print(old_filtered)
new_filtered <- subsetByOverlaps(new_pos, filter)
print(new_filtered)

filtered_aligned <- readGAlignments("~/Projects/netseq-pipeline/test_results/wt-1_20210829/outputs/wt-1.aligned.bam", 
                                    param = ScanBamParam(which = filter))
old_wt2_pos <- import("/n/groups/churchman/GSE159603/GSM4835592_wt-2.pos.bedgraph.gz", genome = "sacCer3")
subsetByOverlaps(old_wt2_pos, filter)bigidx <- (old_wt2_pos$score > 4000)
old_wt2_pos[bigidx]
print(old_wt2_pos[bigidx]$score)
```

```{r}
sessionInfo()
```
