---
title: 'Evaluating a new pipeline for NET-seq data from  GSE159603'
author:
output:  BiocStyle::html_document
params:
  case: ["geo", "new"]
  prefix: ["/n/groups/churchman/GSE159603/", "~/Projects/netseq-pipeline/test_results/wt-1_20210829/outputs/"]
  samples: ["wt-1"]
  tileSize: 1000
  sampleProportion: 0.1
editor_options: 
  markdown: 
    wrap: 72
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(message = FALSE)
suppressPackageStartupMessages({
  library(GenomicRanges)
  library(GenomicAlignments)
  library(rtracklayer)
  
  library(TxDb.Scerevisiae.UCSC.sacCer3.sgdGene)
  library(BSgenome.Scerevisiae.UCSC.sacCer3)
  
  library(kableExtra)
  library(Gviz)
  
  # Tidyverse declared last ... bioconductor homonyms will be masked (e.g. select)
  library(tidyverse)
  library(glue)
  library(plyranges)
  })

set.seed(20210902)
```

## Introduction

The GEO data series GSE159603, "Dynamics of transcription elongation are
finely-tuned by dozens of regulatory factors (baker's yeast)".

Reads-\>Bedgraph

New pipeline. Validate vs GEO results

```{r load.bedgraphs}
strandtab <- c(pos="+", neg="-")
sacCer3Ranges <- SeqinfoForBSGenome("sacCer3")
sacCerRangeNoMito <- dropSeqlevels(sacCer3Ranges, "chrM")

# TODO announce tile size
tiles <- unlist(tileGenome(sacCerRangeNoMito, tilewidth = params$tileSize))
if (params$sampleProportion < 1) {
  tiles <- sort(sample(tiles, length(tiles) * params$sampleProportion))
}

# Create ab, the 'score' data for all the cases
# members are the keys sample, case and strand
# bedgraph - The GRranges from the bedgraph file (with mito chromosome removed)
# rleScore - An Integer Rle of occupancy 
#   (one for each chromosome, see GenomicRanges::moclAsRleList for structure)
ab <- crossing(sample = params$samples, caseid = seq_along(params$case), strand = names(strandtab)) %>% 
  mutate(case = params$case[caseid], dir = params$prefix[caseid],
         name = glue_data(., "{case}.{sample}.{strand}"),
        bedgraph = map(
                      map(glue_data(., "{dir}{sample}.{strand}.bedgraph.gz"), 
                        import.bedGraph, genome = "sacCer3"),
                      dropSeqlevels, 
                        "chrM", pruning.mode = "coarse"),
        occ_tiled = map(ab$bedgraph, function(u) (join_overlap_intersect(u, tiles) %>%
                      mutate(occ = score * width(.))) %>%
                      join_overlap_left(tiles, .) %>%
                      group_by(seqnames, start, end)  %>% 
                      summarize(occ = sum(occ, na.rm = TRUE)))
                    )
###WIP HERE....^^^^^
# TODO Report counts
```

```{r}

tiles <- GenomicRanges::trim(tiles)
tiles$id <- seq_along(tiles)

y <- map(ab$bedgraph, function(u)
    (join_overlap_intersect(u, tiles) %>%
      mutate(occ = score * width(.))) %>%
      join_overlap_left(tiles, .) %>%
      group_by(seqnames, start, end)  %>% 
      summarize(occ = sum(occ, na.rm = TRUE)))

# y <- map(ab$bedgraph, function(u) 
#     (join_overlap_intersect(u, tiles) %>% 
#       mutate(occ = score * width(.))) %>% 
#       as_tibble() %>%
#       group_by(id) %>% 
#       summarize(x = sum(occ)) %>%
#       mutate(chr = seqnames(tiles[id]), start = start(tiles[id]))
#     )
#     
# z <- y %>% mutate(ext  = score * width(.)) %>% group_by(id) %>% summarize(x = sum(occ))

```

```{r}
ab %>% 
  mutate(enframe(lapply(map(ab$rleScore, sum), 
                function(u) u / seqlengths(sacCerRangeNoMito)), 
                name = "idx", value = "chrDensity")) %>%
  select(sample, case, strand, chrDensity) %>%
  unnest_longer(col = chrDensity, indices_to = "seqname") %>%
  mutate(seqname = factor(seqname, levels = seqlevels(sacCerRangeNoMito))) %>%
  relocate(strand, .after = sample) %>%
  relocate(seqname, .after = case) -> occXchr

ggplot(occXchr, aes(x = seqname, y = chrDensity, group = case, col = case)) + geom_point() + facet_wrap(vars(strand), ncol = 1)

occXchr %>% group_by(sample, strand, seqname) %>% spread(seqname, chrDensity) %>% kable(digits = 3)  %>% kable_styling()
  
```



```{r} 
tiled_scores = map(ab$rleScore, 
     function(u) binnedAverage(tiles, u, "meanScore", na.rm=TRUE)$meanScore)

ts <- ab %>% 
  transmute(sample, strand, case, 
            bins = map(ab$rleScore, function(u)
                         binnedAverage(tiles, u, "meanScore", na.rm=TRUE)),
            bin_means = map(bins, function(u) {
                r <- u$meanScore
                names(r) <- seq_along(r)
                r
            })
  ) %>%
  group_by(sample, strand)

# HACK vectors long-to-wide (is there a native way to do this in tidyverse?)
ts_comp <- group_keys(ts) %>% 
  mutate(runs = map(group_split(ts),
                function(u) list(u$bin_means[[1]], u$bin_means[[2]])), 
         geo = runs[[1]], new = runs[[2]])
 
ts_comp %>% mutate(kendalls_tau = 
           map(ts_comp$runs, function(u) 
             cor(u[[1]], u[[2]], method = "kendall")))

ts_data <- ts %>% mutate(bins = NULL) %>%
  group_by(sample, strand, case) %>% unnest_longer(bin_means) %>%
  pivot_wider(id_cols = c(sample, strand), names_from = case, values_from = bin_means)

```

```{r}
v <- data.frame(new = binned_pos_new$meanScore, old = binned_pos_old$meanScore)
plot(v)
filter <- GRanges("chrIV:437850-437860")
 
old_filtered <- subsetByOverlaps(old_pos, filter)
print(old_filtered)
new_filtered <- subsetByOverlaps(new_pos, filter)
print(new_filtered)

filtered_aligned <- readGAlignments("~/Projects/netseq-pipeline/test_results/wt-1_20210829/outputs/wt-1.aligned.bam", 
                                    param = ScanBamParam(which = filter))
old_wt2_pos <- import("/n/groups/churchman/GSE159603/GSM4835592_wt-2.pos.bedgraph.gz", genome = "sacCer3")
subsetByOverlaps(old_wt2_pos, filter)bigidx <- (old_wt2_pos$score > 4000)
old_wt2_pos[bigidx]
print(old_wt2_pos[bigidx]$score)
```

