---
title: 'Evaluating a new pipeline for NET-seq data from  GSE159603'
author:
output:  BiocStyle::html_document
params:
  case: ["geo", "new"]
  prefix: ["/n/groups/churchman/GSE159603/", "~/Projects/netseq-pipeline/test_results/wt-1_20210829/outputs/"]
  samples: ["wt-1"]
  tileSize: 100
  sampleProportion: 0.1
editor_options: 
  markdown: 
    wrap: 72
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(message = FALSE)
suppressPackageStartupMessages({
  library(GenomicRanges)
  library(GenomicAlignments)
  library(rtracklayer)
  
  library(TxDb.Scerevisiae.UCSC.sacCer3.sgdGene)
  library(BSgenome.Scerevisiae.UCSC.sacCer3)
  
  library(kableExtra)
  library(Gviz)
  
  # Tidyverse declared last ... bioconductor homonyms will be masked (e.g. select)
  library(tidyverse)
  library(glue)
  library(plyranges)
  })

set.seed(20210902)
```

## Introduction

The GEO data series GSE159603, "Dynamics of transcription elongation are
finely-tuned by dozens of regulatory factors (baker's yeast)".

Reads-\>Bedgraph

New pipeline. Validate vs GEO results

```{r load.bedgraphs}
strandtab <- c(pos="+", neg="-")
sacCer3Ranges <- SeqinfoForBSGenome("sacCer3")
sacCerRangeNoMito <- dropSeqlevels(sacCer3Ranges, "chrM")

# TODO announce tile size
tiles <- unlist(tileGenome(sacCerRangeNoMito, tilewidth = params$tileSize))
if (params$sampleProportion < 1) {
  tiles <- sort(sample(tiles, length(tiles) * params$sampleProportion))
}

# Create ab, the 'score' data for all the cases
# members are the keys sample, case and strand
# bedgraph - The GRranges from the bedgraph file (with mito chromosome removed)
# rleScore - An Integer Rle of occupancy 
#   (one for each chromosome, see GenomicRanges::moclAsRleList for structure)
ab <- crossing(sample = params$samples, caseid = seq_along(params$case), strand = names(strandtab)) %>%   mutate(case = params$case[caseid], dir = params$prefix[caseid],
         name = glue_data(., "{case}.{sample}.{strand}"),
        bedgraph = map(
                      map(glue_data(., "{dir}{sample}.{strand}.bedgraph.gz"), 
                        import.bedGraph, genome = "sacCer3"),
                      dropSeqlevels, 
                        "chrM", pruning.mode = "coarse")) %>%
        mutate(occ_tiled = map(.$bedgraph, function(u) (join_overlap_intersect(u, tiles) %>%
                  mutate(occ = score * width(.)) %>%
                  join_overlap_left(tiles, .) %>%
                  group_by(seqnames, start, end)  %>% 
                  summarize(occ = sum(occ, na.rm = TRUE), maxocc = replace_na(max(occ) ,0)) %>%
                  as_tibble())
                  )
              )


# TODO Report counts
```

```{r}
o_segs <- ab %>% 
  unnest(occ_tiled) %>% 
  mutate(len = end - start + 1) %>%
  group_by(sample, strand, case, seqnames)

o_tiles <- o_segs %>% 
  summarize(occ = sum(occ, na.rm = TRUE), len = sum(len, na.rm = TRUE)) %>%
  mutate(density = occ / len)

ggplot(o_tiles, aes(x = seqnames, y = density, group = case, col = case)) + geom_point() + facet_wrap(vars(sample, strand), ncol = 1)

o_tiles %>% pivot_wider(id_cols = c(sample, strand, case), names_from = seqnames, values_from = density) %>% kable(digits = 3)  %>% kable_styling()

```

```{r}

dta <- o_segs %>%
  mutate(density = occ / len) %>%
  pivot_wider(id_cols = c(sample, strand, seqnames, start), 
              names_from = case, values_from = density) %>%
  group_by(sample, strand)

dta %>% 
  summarize(tau = cor(geo, new, method = "kendall")) %>% 
  kable(digits = 3) %>% kable_styling()

ggplot(dta, aes(x = geo, y = new)) + 
  geom_point(color = 'blue', alpha = 0.1) + 
  facet_wrap(vars(sample, strand), ncol = 1)  + 
  scale_x_log10() + 
  scale_y_log10()

dtb <- o_segs %>% 
  group_by(sample, strand, case) %>% 
  transmute(sample, strand, case, seqnames, start, density = occ / len)


ggplot(dtb, aes(case, density)) + 
  geom_boxplot(aes(fill = case)) +
  facet_wrap(~sample + strand) +
  scale_y_log10()
```
## What accounts for the density difference?

It looks like it is throughout.

```{r}
q <- ab %>% mutate(q = map(.$bedgraph, 
                      function(u) quantile(rep(u$score, width(u)), c(0,.25,.75,seq(.9,.99,.01),seq(.991,1,.001))))) %>% 
  select(q) %>% 
  unnest_wider(q)

kable(t(q), col.names = ab$name, caption = "Occupancy quantiles") %>% kable_styling()


max_sample_index <- which.max(apply(q, 1, max))
max_score_loc <- (function(u) u[which.max(u$score)])(ab$bedgraph[[max_sample_index]])

```
## Biggest peaks

The highest peak for geo is wildly higher than the higest peaks for new.
The highest peak for wt-1 geo is in sample `r ab$name[max_sample_index]`.
at location `r max_score_loc`.


```{r eval=FALSE}



v <- data.frame(new = binned_pos_new$meanScore, old = binned_pos_old$meanScore)
plot(v)
filter <- GRanges("chrIV:437850-437860")
 
old_filtered <- subsetByOverlaps(old_pos, filter)
print(old_filtered)
new_filtered <- subsetByOverlaps(new_pos, filter)
print(new_filtered)

filtered_aligned <- readGAlignments("~/Projects/netseq-pipeline/test_results/wt-1_20210829/outputs/wt-1.aligned.bam", 
                                    param = ScanBamParam(which = filter))
old_wt2_pos <- import("/n/groups/churchman/GSE159603/GSM4835592_wt-2.pos.bedgraph.gz", genome = "sacCer3")
subsetByOverlaps(old_wt2_pos, filter)bigidx <- (old_wt2_pos$score > 4000)
old_wt2_pos[bigidx]
print(old_wt2_pos[bigidx]$score)
```

```{r}
sessionInfo()
```
